charms/slurmctld/src/interface_influxdb.py

```diff
@@ -0,0 +1,257 @@
+
+
+class InfluxDB(Object):
+    """InfluxDB interface."""
+
+    _on = InfluxDBEvents()
+
+    def __init__(self, charm, relation_name):
+        """Observe relation events."""
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+
+        self._INFLUX_USER = "slurm"
+        self._INFLUX_PRIVILEGE = "all"
+        self._INFLUX_POLICY_NAME = "charmed-hpc"
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_joined,
+            self._on_relation_joined,
+        )
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_broken,
+            self._on_relation_broken,
+        )
+
+    def _on_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Store influxdb_ingress in the charm."""
+        if self.framework.model.unit.is_leader():
+            logger.debug("Slurmctld Leader influxdb._on_relation_joined()")
+
+            if (app_data := event.relation.data.get(self.model.app)) is not None:
+                logger.debug(f"Infulxdb interface app data: {app_data}")
+
+                influxdb_admin_info = app_data.get("influxdb_admin_info", "")
+
+                if influxdb_admin_info != "":
+                    logger.debug(f"Influxdb admin info: {influxdb_admin_info}")
+                    return
+
+                if unit := event.unit:
+                    logger.debug(f"Influxdb unit: {unit}")
+                    if (unit_data := event.relation.data.get(unit)) is not None:
+                        logger.debug(f"Influxdb unit_data: {unit_data}")
+
+                        admin_info = {"host": "", "port": "", "user": "", "password": ""}
+                        if host := unit_data.get("ingress-address"):
+                            admin_info["host"] = host
+                        if port := unit_data.get("port"):
+                            admin_info["port"] = port
+                        if username := unit_data.get("user"):
+                            admin_info["user"] = username
+                        if password := unit_data.get("password"):
+                            admin_info["password"] = password
+
+                        logger.debug(f"Influxdb admin info: {admin_info}")
+
+                        if all(admin_info.values()):
+
+                            app_data["influxdb_admin_info"] = json.dumps(admin_info)
+
+                            # Influxdb client
+                            client = influxdb.InfluxDBClient(
+                                host=admin_info["host"],
+                                port=int(admin_info["port"]),
+                                username=admin_info["user"],
+                                password=admin_info["password"],
+                            )
+
+                            # Influxdb slurm user password
+                            influx_slurm_password = generate_random_string(20)
```

Comment:
Is there a limitation on what characters can be used for the InfluxDB password? If not, could we possibly use `secrets` rather than our own implementation of a random string generator? A 20 character string is relatively easier to brute-force than a 32 character string.

We can use the `token_urlsafe` function from the secrets module to generate passwords for the InfluxDB database:

```pyi
>>> import secrets
>>> secrets.token_urlsafe(32)
'77lIfB5SFHH-GdywGmzSrr9Tp7g9ITYiCYhDv6XMLQE'
```